### PR TITLE
[torch.compile + selective checkpoint] Attach `context_fn` to the checkpointed graph module, fixing flaky tests

### PR DESCRIPTION
**Pull request by yf225**
torch.compile + SAC unit test is causing adjacent unit tests to be flaky due to its modification of shared singleton object. This PR attaches the checkpoint context fn to the checkpointed GraphModule, and look it up during execution, avoiding the need to make the higher-order op stateful.

Specifically, we attach the `context_fn` to the checkpointed GraphModule. These two will be gc'ed at the same time, so it satisfies the lifetime requirement.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by Chillee** [comment]
We're relying on side-tables in a couple different places now in higher order ops, which doesn't seem ... super great to me, particularly if we need to serialize our graphs.

I'm curious what folks think cc: @zou3519 @voznesenskym @ezyang

**Comment by ezyang** [comment]
Is it possible to test this?

File: torch/_higher_order_ops/wrap.py [hunk 1]
```diff
@@ -82,9 +82,10 @@ class TagActivationCheckpoint(HigherOrderOperator):
     in the forward and recomputed forward in backward.
     """
 
+    context_fn_list = []
```
Comment (ezyang) [inline torch/_higher_order_ops/wrap.py:85]: If we're going to be in the business of doing this, this should probably be TLS too

**Review by ezyang — CHANGES_REQUESTED** [review]
`context_fn_list` seems to only ever monotonically grow in size, this does not seem like good design.

However, it's hard for me to give full review here, because I don't really understand what this context fn is doing. Can anyone explain it to me / point me at relevant code comment?

**Comment by zou3519** [comment]
> We're relying on side-tables in a couple different places now in higher order ops, which doesn't seem ... super great to me, particularly if we need to serialize our graphs.
> 
> I'm curious what folks think cc: @zou3519 @voznesenskym @ezyang

Ideally we would serialize it. The way other HOPs do this is to do graph-capture on the function to produce a serializable subgraph, but sometimes we have things that are not really capturable (hooks, triton kernels, the context_fn here).

I'm fine with the sidetable as a stopgap solution.

> However, it's hard for me to give full review here, because I don't really understand what this context fn is doing. Can anyone explain it to me / point me at relevant code comment?

context_fn gets to specify selective checkpointing. It usually returns two torch_dispatch mode.
https://github.com/pytorch/pytorch/blob/2337d8d0625f230f9a0469c5806e282fa4b964e9/torch/utils/checkpoint.py#L1276

File: torch/_dynamo/variables/higher_order_ops.py [hunk 2]
```diff
@@ -1308,9 +1308,11 @@ def call_function(
         from torch.utils.checkpoint import noop_context_fn
         from .builder import wrap_fx_proxy
 
+        context_fn_id = None
         if "context_fn" in kwargs and kwargs["context_fn"] != noop_context_fn:
             context_fn = kwargs.pop("context_fn")
-            self.value.context_fn = context_fn.fn
+            self.value.context_fn_list.append(context_fn.fn)
+            context_fn_id = len(self.value.context_fn_list) - 1
```
Comment (jon-chuang) [inline torch/_dynamo/variables/higher_order_ops.py:1315]: I don't like this. It's a good way to have memory leak bugs. 

Prefer if you can use a hash table (rely on id of wrapper object) and clean up after the lifetime of the higher order op is complete.

File: torch/_higher_order_ops/wrap.py [hunk 3]
```diff
@@ -82,9 +82,16 @@ class TagActivationCheckpoint(HigherOrderOperator):
     in the forward and recomputed forward in backward.
     """
 
+    # Each utils.checkpoint call has an temporary entry in `context_fn_map` storing its context function.
+    # The entry is removed after the checkpointed function is traced through the first time
+    # (at which point the context function is moved to be attached to the graph module).
+    # Reason is that we want the lifetime of the context function to be tied to the graph module
+    # (because as long as the graph module is valid, the context function needs to stay valid),
+    # but there is no way to do this linking during Dynamo tracing, so we have to use this temp map as a bridge.
+    context_fn_tmp_map = {}
```
Comment (yf225) [inline torch/_higher_order_ops/wrap.py:91]: @ezyang I updated it to be a temporary map (with rationale in comment) - wonder should I make this map TLS?

File: torch/_dynamo/variables/higher_order_ops.py (hunk 2 above)
Comment (yf225) [reply torch/_dynamo/variables/higher_order_ops.py]: Please see updated PR - we don't have a good concept of "lifetime of the higher order op", so I attach the context func to the graph module itself. When the graph module is gc'ed, the context func will be gc'ed too.

**Comment by yf225** [comment]
Not ready for review yet, still need some iterations.

**Comment by yf225** [comment]
Updated the PR: we now attach the context_fn to the checkpointed GraphModule instead (these two should always have the same lifetime guarantee). This should solve the lifetime concern.

File: torch/_higher_order_ops/wrap.py (hunk 3 above)
Comment (yf225) [reply torch/_higher_order_ops/wrap.py]: Update: we don't have this global container anymore (and instead we directly attach the context_fn to the checkpointed graph module)

File: torch/_dynamo/variables/higher_order_ops.py (hunk 2 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/higher_order_ops.py]: Hmm, seems better now! 

Can't be completely sure that no two graphmodule can be checkpointed twice? But it seems quite safe as we only add the `_checkpoint_context_fn` meta to the GM that we ourselves create, correct?

It seems like the right object to add this data too. Do you agree @zou3519 ?

File: torch/_higher_order_ops/wrap.py (hunk 3 above)
Comment (jon-chuang) [reply torch/_higher_order_ops/wrap.py]: outdated now right?

File: torch/_dynamo/variables/higher_order_ops.py (hunk 2 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/higher_order_ops.py]: There were some comments about serializing the graph, can the GM meta be serialized? @Chillee

**Review by wanchaol — APPROVED** [review]
The approach looks good to me!

**Pull request closed without merging**